### ISSUE TITLE
Require top-level rules to be nested within an & { ... }

### ISSUE THREAD

**Issue by JacobDB** [stylelint/stylelint#2484]
> Describe the issue. Is it a bug or a feature request (new rule, new option, etc.)?

Feature request

> Which rule, if any, is this issue related to?

Possibly `selector-nested-pattern`

> What CSS is needed to reproduce this issue?

Invalid:

```css
.foo {
    color: pink;
}
````

Valid:

```css
.foo {
    & {
        color: pink;
    }
}
```

> What stylelint configuration is needed to reproduce this issue?

N/A

> Which version of stylelint are you using?

`7.10.1`

> How are you running stylelint: CLI, PostCSS plugin, Node API?

Node API, with gulp-stylelint

> Does your issue relate to non-standard syntax (e.g. SCSS, nesting, etc.)?

Yes, it's related to nesting.

> What did you expect to happen?

A warning to be flagged when parsing a top-level rule that has properties that are not nested. Some examples:

Invalid:

```css
.foo {
    color: pink;
}

.foo {
    color: pink;

    &.-valid {
        color: teal;
    }

    .bar {
        color: red;
    }
}
```

Valid:

```css
.foo {
    & {
        color: pink;
    }
}

.foo {
    & {
        color: pink;
    }

    &.-variant {
        color: teal;
    }

    .bar {
        color: red;
    }
}
```

> What actually happened (e.g. what warnings or errors you are getting)?

To my knowledge, no rule exists that can accomplish this.

**Comment by evilebottnawi** [comment]
@JacobDB in stylelint-scss we have similar rule https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/selector-no-redundant-nesting-selector/README.md

**Comment by JacobDB** [comment]
@evilebottnawi I saw that while researching, but isn't that kind of the opposite of what I'm looking to do?

**Comment by jeddy3** [comment]
@JacobDB Thanks for the suggestion and for using the issue template.

> To my knowledge, no rule exists that can accomplish this.

Yes, you are right, no rules exists for this.

I think this is a great candidate for a [plugin](https://stylelint.io/developer-guide/plugins/). What you're after is kind of specific, and so I'm not quite sure what you'd call your plugin... maybe, `stylelint-rule-no-top-level-properties` or something.

**Comment by evilebottnawi** [comment]
@jeddy3 i think this is a good candidate for the core. `cssnext` is coming ... :snowflake:

**Comment by jeddy3** [comment]
It's easier for this to start as a plugin, and then migrate it to core if suitable, than start it in core and then have to deprecate it.

**Comment by evilebottnawi** [comment]
@jeddy3 you are right, let's start as plugin, after nesting accepted as recommended spec we can migrate in the core. In fact, I initially thought so, just leave this comment if someone wants to understand our position.

**Comment by jeddy3** [comment]
@JacobDB I'm going to label this as a plugin and close. We've [a guide to writing plugins](https://stylelint.io/developer-guide/plugins/). Please drop us a PR adding the plugin to the [plugin list](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/plugins.md), if you do write a plugin. Good luck!

**Comment by JacobDB** [comment]
@jeddy3 sure, thanks for the info. I'll try and take a crack at it eventually, will report back if I have any success.
